// livereactload/lib/browserify-plugin/server.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.startServer = startServer;

var _ws = require("ws");

var _console = require("./console");

var _https = require("https");

var _https2 = _interopRequireDefault(_https);

var _fs = require("fs");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function logError(error) {
  if (error) {
    (0, _console.log)(error);
  }
}

function startServer(_ref) {
  var port = _ref.port,
      sslKey = _ref.sslKey,
      sslCert = _ref.sslCert;

  if (sslCert && !sslKey || !sslCert && sslKey) {
    throw new Error('You need both a certificate AND key in order to use SSL');
  }

  var wss = void 0;
  if (sslCert && sslKey) {
    var key = (0, _fs.readFileSync)(sslKey, 'utf8');
    var cert = (0, _fs.readFileSync)(sslCert, 'utf8');
    var credentials = { key: key, cert: cert };
    var _server = _https2.default.createServer(credentials);
    _server.listen(port);
    wss = new _ws.Server({ server: _server });
  } else {
    wss = new _ws.Server({ port: port });
  }

  (0, _console.log)("Reload server up and listening in port " + port + "...");

  var server = {
    notifyReload: function notifyReload(metadata) {
      if (wss.clients.length) {
        (0, _console.log)("Notify clients about bundle change...");
      }
      wss.clients.forEach(function (client) {
        client.send(JSON.stringify({
          type: "change",
          data: metadata
        }), logError);
      });
    },
    notifyBundleError: function notifyBundleError(error) {
      if (wss.clients.length) {
        (0, _console.log)("Notify clients about bundle error...");
      }
      wss.clients.forEach(function (client) {
        client.send(JSON.stringify({
          type: "bundle_error",
          data: { error: error.toString() }
        }), logError);
      });
    }
  };

  wss.on("connection", function (client) {
    (0, _console.log)("New client connected");
  });

  return server;
}
